source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h

```diff
@@ -237,35 +240,137 @@ class UdpProxyFilter : public Network::UdpListenerReadFilter,
     }
   };
 
+  struct LocalPeerHostAddresses {
+    const Network::UdpRecvData::LocalPeerAddresses& local_peer_addresses_;
+    const Upstream::Host& host_;
+  };
+
+  struct HeterogeneousActiveSessionHashWithUpstreamHost {
```

Comment:
Couldn't all of this code be unified further by just always using the upstream host included version but setting it to null in the case of sticky sessions? Conversely, in the per datagram load balancing case, why does peer address matter at all? It seems like basically we either want to hash only on host or only on peer address? And I'm wondering if it could all be unified further?

/wait

Reply:
I think that the former option is a good one (having null host). This will also simplify code significantly.
On you latter approach:
I think it's good to have peer, local and upstream host taken into consideration because:
- if we take only upstream host, other clients may reuse the same UDP socket if upstream host matches. I don't think that's a good idea
- if we take peer only, we'll treat UDP sockets for that peer to different hosts as duplicates and therefore each will be removed after creation of a new one. On my opinion, possibility of reusing already created UDP sockets for multiple host for each client may be beneficial if we have many datagrams and a few upstream hosts.